src/data_device_manager/data_device.rs

```diff
@@ -0,0 +1,218 @@
+use std::sync::Mutex;
+
+use wayland_client::{
+    event_created_child,
+    protocol::{
+        wl_data_device::{self, WlDataDevice},
+        wl_data_offer::{self, WlDataOffer},
+        wl_surface::WlSurface,
+    },
+    Connection, Dispatch, QueueHandle,
+};
+
+use super::{
+    data_offer::{DataOfferData, DataOfferHandler},
+    DataDeviceManagerState,
+};
+
+#[derive(Debug, Default)]
+pub struct DataDeviceInner {
+    /// the active DnD data offer and its data
+    pub dnd_data_offer: Option<DndDataOffer>,
+    /// the active selection data offer and its serial
+    pub selection: Option<WlDataOffer>,
+}
+
+#[derive(Debug, Clone)]
+pub struct DndDataOffer {
+    /// the wl_data offer if it exists
+    pub data_offer: Option<WlDataOffer>,
+    /// the serial for this data offer
+    pub serial: u32,
+    /// the surface that this DnD is active on
+    pub surface: WlSurface,
+    /// the x position on the surface
+    pub x: f64,
+    /// the y position on this surface
+    pub y: f64,
+    /// the timestamp a motion event was received in millisecond granularity
+    pub time: Option<u32>,
+}
+
+#[derive(Debug, Default)]
+pub struct DataDeviceData {
+    pub(super) inner: Mutex<DataDeviceInner>,
+}
+
+pub trait DataDeviceDataExt: Send + Sync {
+    fn data_device_data(&self) -> &DataDeviceData;
+
+    /// get the active dnd offer if it exists
+    fn dnd_offer(&self) -> Option<DndDataOffer> {
+        self.data_device_data().inner.lock().ok().and_then(|inner| inner.dnd_data_offer.clone())
+    }
+
+    /// get the active selection offer if it exists
+    fn selection_offer(&self) -> Option<WlDataOffer> {
+        self.data_device_data().inner.lock().ok().and_then(|inner| inner.selection.clone())
+    }
```

Comment:
These two functions are override-able to  handle dnd and selection offers that an extension may use I assume